### PR TITLE
allow rspec 3

### PR DESCRIPTION
**Pull request by pvmeerbe**
modified gem to support rspec 3:

- changed gemspec dependency
- updated spec tests : remove deprecated warnings by using the  new 'expect' syntax instead of 'should' 
- converted  HTMLFormatter to new  RSpec 3 format

### CODE REVIEW

File: lib/watir/rspec/helper.rb [hunk 1]
```diff
@@ -21,10 +21,19 @@ def method_missing(name, *args)
         if browser.respond_to?(name)
           Helper.module_eval %Q[
             def #{name}(*args)
-              browser.send(:#{name}, *args) {yield}
+              if block_given?
```
Comment (jarmo) [inline lib/watir/rspec/helper.rb]: Why is this change necessary? Did something break?

File: lib/watir/rspec/version.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 module Watir
   class RSpec
-    VERSION = "1.1.2"
+    VERSION = "1.1.4"
```
Comment (jarmo) [inline lib/watir/rspec/version.rb]: Please do not change version in pull requests - i'll do it myself when it's time to release. I'm also trying to follow [Semantic Versioning](http://semver.org) and these changes include backward incompatible changes which mean that the major version should probably be increased.

File: lib/watir/rspec/cli.rb [hunk 3]
```diff
@@ -88,6 +88,8 @@ def install
 end
   ]
           end
+
+          puts "WARNING : the HTML formatter cannot be used until the rspec legacy formatter gem is also installed !! (https://github.com/rspec/rspec-legacy_formatters)"
```
Comment (jarmo) [inline lib/watir/rspec/cli.rb]: Since we are already forcing to use `RSpec 3` then why not include that legacy formatter gem as a `watir-rspec` dependency as long as formatter hasn't updated to `RSpec 3` as well? Do you see any problems by doing like that?

File: lib/watir/rspec/helper.rb (hunk 1 above)
Comment (pvmeerbe) [inline lib/watir/rspec/helper.rb]: I use a custom layer between rspec & watir/selenium which optionally allows block passing on several location. IMHO the original code will always yield a block even if the original call catched  by the method_missing was not, causing weird behavior in my code. The patch will only pass a block if requested.

File: spec/watir/rspec/helper_spec.rb [hunk 4]
```diff
@@ -7,50 +7,52 @@
...
     it "redirects missing methods to browser if method exists" do
       @browser = double("browser", coolness_factor: :very)
-      coolness_factor.should == :very
+      expect(coolness_factor).to eql :very
     end
 
     it "raises error when browser does not have method" do
       @browser = double("browser")
-      described_class.should_not be_method_defined :not_existing_method
+      expect(described_class.instance_methods).not_to include :method_to_be_defined
```
Comment (jarmo) [inline spec/watir/rspec/helper_spec.rb]: You've changed here previous behavior of the test by changing that method name. Why did you do that?

File: spec/watir/rspec/helper_spec.rb [hunk 5]
```diff
@@ -7,50 +7,52 @@
...
     end
 
     it "adds browser methods to the helper" do
-      @browser = double("browser", method_to_be_defined: :done)
-      described_class.should_not be_method_defined :method_to_be_defined
-      method_to_be_defined.should == :done
-      described_class.should be_method_defined :method_to_be_defined
+      @browser = double("browser", method_to_be_defined2: :done)
```
Comment (jarmo) [inline spec/watir/rspec/helper_spec.rb]: Why rename the method?

File: spec/watir/rspec/matchers/base_matcher_spec.rb [hunk 6]
```diff
@@ -134,24 +134,24 @@
         end
       end
 
-      object.should be_foo.during(0.1)
-      raised.should be_true
+      expect(object).to be_foo.during(0.1)
+      expect(raised).to be_truthy
```
Comment (jarmo) [inline spec/watir/rspec/matchers/base_matcher_spec.rb]: Please use `be_true` here since we care in this test that it is actually `true` and not anything else, because anything else might mean that functionality is not working correctly.

File: spec/watir/rspec/matchers/base_matcher_spec.rb [hunk 7]
```diff
@@ -96,8 +96,8 @@
         end
       end
 
-      object.should_not be_foo.within(1)
-      raised.should be_true
+      expect(object).not_to be_foo.within(1)
+      expect(raised).to be_truthy
```
Comment (jarmo) [inline spec/watir/rspec/matchers/base_matcher_spec.rb]: Please use be_true here since we care in this test that it is actually true and not anything else, because anything else might mean that functionality is not working correctly.

File: spec/watir/rspec/matchers/base_matcher_spec.rb [hunk 8]
```diff
@@ -61,17 +61,17 @@
         end
       end
 
-      object.should be_foo.within(1)
-      raised.should be_true
+      expect(object).to be_foo.within(1)
+      expect(raised).to be_truthy
```
Comment (jarmo) [inline spec/watir/rspec/matchers/base_matcher_spec.rb]: Please use be_true here since we care in this test that it is actually true and not anything else, because anything else might mean that functionality is not working correctly.

File: spec/watir/rspec/matchers/base_matcher_spec.rb [hunk 9]
```diff
@@ -168,14 +168,14 @@
         end
       end
 
-      object.should_not be_foo.during(0.1)
-      raised.should be_true
+      expect(object).not_to be_foo.during(0.1)
+      expect(raised).to be_truthy
```
Comment (jarmo) [inline spec/watir/rspec/matchers/base_matcher_spec.rb]: Please use be_true here since we care in this test that it is actually true and not anything else, because anything else might mean that functionality is not working correctly.

File: lib/watir/rspec/cli.rb (hunk 3 above)
Comment (pvmeerbe) [inline lib/watir/rspec/cli.rb]: that is indeed an option.

however I've noticed that even with the legacy gem there is still an error when a test fails:

NoMethodError: undefined method `before_all_ivars' for RSpec::ExampleGroups::GatekeeperEvalLicenses:Class
    occurred at /rspec-core-3.2.0/lib/rspec/core/example_group.rb:634:in `method_missing'


The before_all_ivars method was indeed removed in rspec-core ;( So I guess the best solution is to rewrite the HTML formatter using the new syntax

File: watir-rspec.gemspec [hunk 10]
```diff
@@ -15,7 +15,7 @@ Gem::Specification.new do |gem|
   gem.require_paths = ["lib"]
   gem.version       = Watir::RSpec::VERSION
 
-  gem.add_dependency "rspec", "~>2.0"
+  gem.add_dependency "rspec", "~>3.0"
```
Comment (jarmo) [inline watir-rspec.gemspec]: Now, to think about it - do we really need it to increase to `3.0`? Do we have currently anything `RSpec 3` specific, which is not working in `RSpec 2`? Isn't `expect` syntax also working with `RSpec 2`? Not sure and i don't have anything against increasing the version number here, but just made me think that do we actually need to do that.

File: lib/watir/rspec/version.rb (hunk 2 above)
Comment (pvmeerbe) [inline lib/watir/rspec/version.rb]: reverted version change done

File: spec/watir/rspec/helper_spec.rb (hunk 5 above)
Comment (pvmeerbe) [inline spec/watir/rspec/helper_spec.rb]: I'm having issues with this test when using the 

expect(described_class).to respond_to :method_to_be_defined
  and
expect(described_class).not_to respond_to :method_to_be_defined

somehow the last two tests seem to interfere unless I use the syntax as commited  + use a different method name for the two tests. I don't really understand why yet ....
I'll see if I can digg in to this next Tuesday .

Comment (jarmo) [inline spec/watir/rspec/helper_spec.rb]: Is it possible that `RSpec` magic helper breaks it? In that case, just revert back to using `instance_methods` and call it a day :) Sometimes RSpec tries to be too smart and while doing that leaves a mess behind.

File: lib/watir/rspec/html_formatter.rb [hunk 11]
```diff
@@ -38,7 +38,8 @@ def example_started(example)
 
       # @private
       def extra_failure_content(exception)
-        browser = example_group.before_all_ivars[:@browser] || $browser
+        return super unless example_group  # apparently there are cases where rspec failures are encountered and the example_group is not set (i.e. nil)
```
Comment (jarmo) [inline lib/watir/rspec/html_formatter.rb]: What? Really? Why is reporter invoked in the first place then? :D

File: spec/watir/rspec/helper_spec.rb [hunk 12]
```diff
@@ -7,50 +7,53 @@
...
-      described_class.should_not be_method_defined :not_existing_method
+      expect(described_class).not_to respond_to :not_existing_method
     end
 
     it "adds browser methods to the helper" do
-      @browser = double("browser", method_to_be_defined: :done)
-      described_class.should_not be_method_defined :method_to_be_defined
-      method_to_be_defined.should == :done
-      described_class.should be_method_defined :method_to_be_defined
+      @browser = double("browser", not_existing_method2: :done)
```
Comment (jarmo) [inline spec/watir/rspec/helper_spec.rb]: Please use same method name as before, since it's more clear. Also, remove comments.

**Comment by jarmo** [comment]
What's the state of this PR? I've added some additional comments, but can see that you've made additional commits as well. Please, if adding other commits not directly related with RSpec 3, use a different branch or make a PR with a separate branch and close this one.

File: watir-rspec.gemspec (hunk 10 above)
Comment (pvmeerbe) [inline watir-rspec.gemspec]: I didn't  try running the new version with RSpec 2.X.
However I assume it will not work as  the HTMLFormatter is now adjusted to work with RSpec v 3 (it uses a ::RSpec::Core::Formatters.register which is not available in version 2.99)

File: lib/watir/rspec/cli.rb (hunk 3 above)
Comment (pvmeerbe) [inline lib/watir/rspec/cli.rb]: I converted the HTML formatter to rspec version 3 , so the legacy gem is no longer needed

File: spec/watir/rspec/helper_spec.rb (hunk 5 above)
Comment (pvmeerbe) [inline spec/watir/rspec/helper_spec.rb]: I kept instance_method as somehow there seems to be some interference...

File: spec/watir/rspec/helper_spec.rb (hunk 4 above)
Comment (pvmeerbe) [inline spec/watir/rspec/helper_spec.rb]: see related comment below

File: spec/watir/rspec/matchers/base_matcher_spec.rb (hunk 6 above)
Comment (pvmeerbe) [inline spec/watir/rspec/matchers/base_matcher_spec.rb]: done

File: spec/watir/rspec/matchers/base_matcher_spec.rb (hunk 7 above)
Comment (pvmeerbe) [inline spec/watir/rspec/matchers/base_matcher_spec.rb]: done

File: spec/watir/rspec/matchers/base_matcher_spec.rb (hunk 8 above)
Comment (pvmeerbe) [inline spec/watir/rspec/matchers/base_matcher_spec.rb]: done

File: spec/watir/rspec/matchers/base_matcher_spec.rb (hunk 9 above)
Comment (pvmeerbe) [inline spec/watir/rspec/matchers/base_matcher_spec.rb]: done

**Comment by pvmeerbe** [comment]
sorry jarmo, I did not see your comment from 5 days ago & already made additional changes/comments. I updated the original the original PR description to include all changes made.

IMHO all your change requests are done   +  everything seems to be working now (spec tests are green + then  new version works in my environment)

So I'm done  unless you have additional change requests

File: spec/watir/rspec/helper_spec.rb (hunk 5 above)
Comment (jarmo) [inline spec/watir/rspec/helper_spec.rb]: There shouldn't be any. Or if there is then something seems to be broken...

Comment (jarmo) [inline spec/watir/rspec/helper_spec.rb]: Is it possible that the culprit is previous test where you've also renamed method `:not_existing_method`? Please rename both back to original names.

File: watir-rspec.gemspec (hunk 10 above)
Comment (jarmo) [inline watir-rspec.gemspec]: That wasn't before the case. I'm not familiar with RSpec 3 formatters, but is that registering really needed? Can't you just configure to use custom formatter like with RSpec 2? What does that registering mean in context of RSpec?

File: spec/watir/rspec/helper_spec.rb (hunk 5 above)
Comment (pvmeerbe) [inline spec/watir/rspec/helper_spec.rb]: found my issue ; I used respond_to instead of the original method_defined

Since it is used on a module these behave differently  :§. I reused be_method_defined & the original method names and now it works fine.

File: watir-rspec.gemspec (hunk 10 above)
Comment (pvmeerbe) [inline watir-rspec.gemspec]: See https://github.com/rspec/rspec-core/blob/master/lib/rspec/core/formatters.rb 
The register code is  @ line 81

If you don't register you get the warning on deprecated formatter and the formatteri isn't used : see line 149

There isn't much documentation on this, but as far as I can see there are 2 options:
* register your formatter 
* use the rspec-legacy_formatter gem & the original formatter (you have to rename the before_all_ivars method to before_context_ivars;  see commit https://github.com/rspec/rspec-core/commit/639db76e5fd0e845ae6c44527a0ba750cb934acd  example_group.rb file)

**Comment by jarmo** [comment]
Ran specs on my machine and still having two deprecation warnings:
```
Deprecation Warnings:

Using `should` from rspec-expectations' old `:should` syntax without explicitly enabling the syntax is deprecated. Use the new `:expect` syntax or explicitly enable `:should` with `config.expect_with(:rspec) { |c| c.syntax = :should }` instead. Called from watir-rspec/spec/watir/rspec_spec.rb:18:in `block (3 levels) in <top (required)>'.

Using `stub` from rspec-mocks' old `:should` syntax without explicitly enabling the syntax is deprecated. Use the new `:expect` syntax or explicitly enable `:should` instead. Called from watir-rspec/spec/watir/rspec_spec.rb:15:in `block (3 levels) in <top (required)>'.


If you need more of the backtrace for any of these deprecations to
identify where to make the necessary changes, you can configure
`config.raise_errors_for_deprecations!`, and it will turn the
deprecation warnings into errors, giving you the full backtrace.

2 deprecation warnings total
```

Also, does this code work with RSpec 3 as well? https://github.com/watir/watir-rspec/blob/master/lib/watir/rspec.rb#L24-L39

I'm not sure if it's being hit by specs because it would need specs to actually load that `HtmlFormatter`.

**Comment by pvmeerbe** [comment]
The 2 deprecation warnings come from this commit https://github.com/watir/watir-rspec/commit/d7272c6d9a6fa269019ae2611eb05cee2ddc8e10#diff-c35e68030088f30c29b8ccccc114c0de added after my fork.

I rebased my fork & converted the feature.

Regarding https://github.com/watir/watir-rspec/blob/master/lib/watir/rspec.rb#L24-L39 code:  I tried calling Watir::RSpec.file_path("foobar")  when formatter is configured in the spec_helper.rb file & without configuring it (in which case the error is raised). Both scenarios seem to work.

**Comment by jarmo** [comment]
Sorry about that.

Finally merged. Thanks!

One thing i'm still missing is a spec for functionality created at https://github.com/watir/watir-rspec/blob/master/lib/watir/rspec/helper.rb#L32-L34 - i bet that's the reason why code coverage decreased as well.

Also, could you show some real-life code which uses that feature, since i can't recall any `Watir::Browser` methods right now, which might take a block as an argument.

**Comment by pvmeerbe** [comment]
I'm using a custom layer around Watir with several block methods for example to run a code block after visiting a page. 

But a quick look at the API of watir shows http://www.rubydoc.info/gems/watir-webdriver/Watir/AlertHelper which has several block methods. Or the use method of Window : http://www.rubydoc.info/gems/watir-webdriver/Watir/Window:use


Thanks for merging & releasing!

**Comment by jarmo** [comment]
`AlertHelper` seems to be a deprecated API as written in the docs.

`Window#use` does indeed accept block, but not `Browser#window`, which means that everything would have worked without the changes related to blocks.

But reading at your original problem then did you know that `Watir::Browser` has an API for that - http://www.rubydoc.info/gems/watir-webdriver/Watir/Browser#add_checker-instance_method and by looking at that you might need to provide a block.
